### ISSUE TITLE
Guidance needed on adding dynamic context - mixin / formatters / serializers

### ISSUE THREAD

**Issue by KidkArolis** [pinojs/pino#867]
Hi.. first of all, love pino, been a very helpful solid library so far.

However, I keep coming back to trying to find the best way to add dynamic context into log messages _at log time_ (e.g. current user, and some other meta data).

More concretely, if I have a set of middleware processing a request, the request might get gradually extended, e.g.:

1. request is made
2. request id is generated
3. user is authenticated
4. some params are calculated

If I log something after each of those steps, I want to log as much context as is already available, e.g. if user is present on req.user - log it, otherwise don't. That kind of thing (I donno, maybe this is bad design?).

And it seems like there's no clear way of doing this, I tried the following.

## Mixin

This doesn't work - the mixin is never called on child loggers (?).

```js
function createRequestLogger(context) {
  return context.app.get('logger').child({
    mixin() {
      return calculateDynamicMetaFromContext(context)
    }
  })
}
```

## Serializers

This doesn't work - the result of the serializer is cached and is incorrect upon later logs.

```js
function createRequestLogger(context) {
  return context.app.get('logger').child({
    context,
    serializers: {
      context: calculateDynamicMetaFromContext
    }
  })
}
```

## Formatters

This kind of works - but formatter is being called twice during a log call (See #866) and can yield incorrect results if the formatter _removes_ something from the provided object, because the 2nd time it's called that extra meta is now been deleted.

```js
function createRequestLogger(context) {
  return context.app.get('logger').child({
    formatters: {
      log: calculateDynamicMetaFromContext(context)
    }
  })
}
```

In other words - I don't know why there are 3 ways to do something very similar yet all ways are strangely confusing with their own quirks 🤷‍♂️

**Comment by jsumners** [comment]
A child logger is used to create a new logger instance that inherits all of its parent's metadata (e.g. logger name) and attach new metadata for all logs written by that child logger. In general, the usage pattern is something like:

```js
const parent = pino({ name: 'silly-service' })

// do some stuff
parent.info('parent has done some stuff')

function onIncomingRequest (req, res, next) {
  req.log = parent.child({ requestId: 'req123' })
  next()
}

function requestHandler (req, res) {
  // this is the child logger created in onIncomingRequest
  req.log({ user: req.user, something: 'else' }, 'handling request')

  res.send({ hello: 'world' })
}
```

**Comment by KidkArolis** [comment]
That's a great example to build upon! In your example, you're adding the user to the context by passing it into the log call. So the next question (my original question, in fact) would be - how do you inject the user into the logger's context such that individual log calls don't have to manually pass the user in.

I tried to reproduce each of the scenarios above with that use case in mind and show why neither works (except the last scenario sort of works):

## Mixin

Issue: mixin is never called and therefore user is never logged.

```js
const pino = require('pino')

const logger = pino()

const context = {}

const child = logger.child({
  mixin() {
    console.log('Calling mixin?')
    return { user: context.user }
  }
})

child.info('Logging before user is available')
context.user = { id: 1 }
child.info('Logging after user is available')

// Output: no user provided

// {"level":30,"time":1593173891224,"pid":7109,"hostname":"inrainbows.local","msg":"Logging before user is available"}
// {"level":30,"time":1593173891224,"pid":7109,"hostname":"inrainbows.local","msg":"Logging after user is available"}
```

## Serializer

Issue: serializer is cached and therefore user is never logged.

```js
const pino = require('pino')

const logger = pino()

const context = {}

const child = logger.child({
  context,
  serializers: {
    context: (context) => {
      console.log('Calling serializer?')
      return { user: context.user }
    }
  }
})

child.info('Logging before user is available')
context.user = { id: 1 }
child.info('Logging after user is available')

// Output: no user provided
// 
// Calling serializer?
// {"level":30,"time":1593174010110,"pid":7129,"hostname":"inrainbows.local","context":{},"msg":"Logging before user is available"}
// {"level":30,"time":1593174010111,"pid":7129,"hostname":"inrainbows.local","context":{},"msg":"Logging after user is available"}
```

## Formatter

Ok, this works!

**However**, if you turn on "prettyPrint: true" when creating pino (which we use in development), the code below throws on this line. Which means it's not possible for a formatter to "rewrite" the message on the fly in such a way, which admittedly can be avoided by not deleting the object in the first place. But I'm not sure how to "reason" about this? This is the issue flagged in #866.

```
obj.responseHeaders = obj.res.headers
                                    ^

TypeError: Cannot read property 'headers' of undefined
```

```js
const pino = require('pino')

const logger = pino()

const context = {}

const child = logger.child({
  context,
  formatters: {
    log: (obj) => {
      obj.user = context.user
      obj.responseHeaders = obj.res.headers
      delete obj.res
      return obj
    }
  }
})

child.info({ res: { headers: 'x' } }, 'Logging before user is available')
context.user = { id: 1 }
child.info({ res: { headers: 'x' } }, 'Logging after user is available')

// Output: no user provided
// 
// Calling serializer?
// {"level":30,"time":1593174188579,"pid":7160,"hostname":"inrainbows.local","context":{},"responseHeaders":"x","msg":"Logging before user is available"}
// {"level":30,"time":1593174188580,"pid":7160,"hostname":"inrainbows.local","context":{},"user":{"id":1},"responseHeaders":"x","msg":"Logging after user is available"}
```

**Comment by jsumners** [comment]
I don't really know anything about the `mixin` feature. @watson added it. Maybe he can help there?

Yes, serializers are executed on child creation via

https://github.com/pinojs/pino/blob/faf7cb85fe645cfec08a12b46ea0dc21c0f77f11/lib/tools.js#L169

However, you also have a mistake in your example. You are not providing anything for the logger to serialize.

```js
context.user = { id: 1 }
child.info('Logging after user is available')
// ^ should be ...
child.info({ context}, 'Logging after user is available')
```

As for pretty printing breaking things, you might have found a bug. Would you like to submit a PR to resolve it?

**Comment by KidkArolis** [comment]
> However, you also have a mistake in your example. You are not providing anything for the logger to serialize.

I am passing context in the child creation:

![image](https://user-images.githubusercontent.com/324440/85858814-27537a80-b7b4-11ea-8fe7-de810ee727e7.png)

**Comment by KidkArolis** [comment]
So apart from the formatter ~bug, would you say formatters is the best way to add dynamic context to log messages - if so - should `mixin` be removed altogether as it seems to be superseded by formatters?

**Comment by jsumners** [comment]
No. I would say to write logs with the context. You don't need any extra features for it:

```js
log.info({ context: 'object' })
```

**Comment by KidkArolis** [comment]
@jsumners when would you say child logger bindings should be used then if ever?

**Comment by jsumners** [comment]
For things like http request metadata, e.g. `requestId`. Or if you need to pass a logger into nested dependencies, e.g.:

**index.js**
```js
const log = require('pino')()
const depA = require('depA')({ logger: log })
```

**depA**
```js
module.exports = function ({ logger }) {
  this.log = logger.child({ mod: 'depA' })
  const depB = require('depB')({ logger: this.log })

  return this
}
```

Child logger metadata is meant to be _static_ metadata. Point-in-time metadata should be logged using the logger methods as described previously and in the documentation.
